diff --git a/go.env b/go.env
index 91f1490c15cda3..f7d6007211461d 100644
--- a/go.env
+++ b/go.env
@@ -7,6 +7,6 @@
 GOPROXY=https://proxy.golang.org,direct
 GOSUMDB=sum.golang.org
 
-# Automatically download newer toolchains as directed by go.mod files.
+# Use the locally installed Go toolchain, never downloading a different one.
 # See https://go.dev/doc/toolchain for details.
-GOTOOLCHAIN=auto
+GOTOOLCHAIN=local
diff --git a/src/cmd/link/internal/ld/pe.go b/src/cmd/link/internal/ld/pe.go
index 8376ddefe56bb7..bd12916d4e9dcb 100644
--- a/src/cmd/link/internal/ld/pe.go
+++ b/src/cmd/link/internal/ld/pe.go
@@ -282,7 +282,7 @@ type IMAGE_LOAD_CONFIG_DIRECTORY64 struct {
 }
 
 const (
-	PeMinimumTargetMajorVersion = 10
+	PeMinimumTargetMajorVersion = 6
 	PeMinimumTargetMinorVersion = 0
 )
 
diff --git a/src/runtime/os_windows.go b/src/runtime/os_windows.go
index 42af4200547b64..209b40547ee786 100644
--- a/src/runtime/os_windows.go
+++ b/src/runtime/os_windows.go
@@ -137,7 +137,6 @@ var (
 	_NtCreateWaitCompletionPacket    stdFunction
 	_NtAssociateWaitCompletionPacket stdFunction
 	_NtCancelWaitCompletionPacket    stdFunction
-	_RtlGetCurrentPeb                stdFunction
 	_RtlGetVersion                   stdFunction
 
 	// These are from non-kernel32.dll, so we prefer to LoadLibraryEx them.
@@ -292,7 +291,6 @@ func loadOptionalSyscalls() {
 			throw("NtCreateWaitCompletionPacket exists but NtCancelWaitCompletionPacket does not")
 		}
 	}
-	_RtlGetCurrentPeb = windowsFindfunc(n32, []byte("RtlGetCurrentPeb\000"))
 	_RtlGetVersion = windowsFindfunc(n32, []byte("RtlGetVersion\000"))
 }
 
@@ -444,26 +442,7 @@ var canUseLongPaths bool
 
 // initLongPathSupport enables long path support.
 func initLongPathSupport() {
-	const (
-		IsLongPathAwareProcess = 0x80
-		PebBitFieldOffset      = 3
-	)
-
-	// Check that we're ≥ 10.0.15063.
-	info := windows.OSVERSIONINFOW{}
-	info.OSVersionInfoSize = uint32(unsafe.Sizeof(info))
-	stdcall(_RtlGetVersion, uintptr(unsafe.Pointer(&info)))
-	if info.MajorVersion < 10 || (info.MajorVersion == 10 && info.MinorVersion == 0 && info.BuildNumber < 15063) {
-		return
-	}
-
-	// Set the IsLongPathAwareProcess flag of the PEB's bit field.
-	// This flag is not documented, but it's known to be used
-	// by Windows to enable long path support.
-	bitField := (*byte)(unsafe.Pointer(stdcall(_RtlGetCurrentPeb) + PebBitFieldOffset))
-	*bitField |= IsLongPathAwareProcess
-
-	canUseLongPaths = true
+	canUseLongPaths = false
 }
 
 func osinit() {
